src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+   * @return The GMC/GDC number, depending on which is valid.
+   */
+  private String getReferenceNumber(PersonalDetails personalDetails) {
+    String gmcNumber = personalDetails.getGmcNumber();
+    return gmcNumber.matches("\\d{7}") ? gmcNumber : personalDetails.getGdcNumber();
+  }
+
+  /**
+   * Get the NTN suffix for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to get the suffix for.
+   * @return The calculated suffix for the programme membership's NTN.
+   */
+  private String getSuffix(ProgrammeMembership programmeMembership) {
+    log.info("Calculating suffix.");
+    String trainingPathway = programmeMembership.getTrainingPathway();
+    log.info("Using training pathway '{}' to calculate suffix.", trainingPathway);
+
+    String suffix = switch (trainingPathway) {
+      case "CCT" -> "C";
+      case "CESR" -> "CP";
+      default -> {
+        List<Curriculum> curricula = programmeMembership.getCurricula();
+        List<Curriculum> sortedCurricula = filterAndSortCurricula(curricula);
+        String firstSpecialtyCode = sortedCurricula.get(0).getCurriculumSpecialtyCode();
+        log.info("Using specialty code '{}' to calculate suffix.", trainingPathway);
+
+        yield Objects.equals(firstSpecialtyCode, "ACA") ? "C" : "D";
+      }
+    };
+
+    log.info("Calculated suffix: '{}'.", suffix);
+    return suffix;
+  }
+
+  /**
+   * Filter the given curricula to only those that are current and sort them alphanumerically.
+   *
+   * @param curricula The curricula to filter and sort.
+   * @return The current curricula, sorted alphanumerically by subtype and code.
+   */
+  private List<Curriculum> filterAndSortCurricula(List<Curriculum> curricula) {
+    LocalDate now = LocalDate.now();
+
+    return curricula.stream()
+        .filter(c ->
+            c.getCurriculumSpecialtyCode() != null && !c.getCurriculumSpecialtyCode().isBlank())
+        .filter(c -> c.getCurriculumStartDate().isBefore(now))
+        .filter(c -> c.getCurriculumEndDate().isAfter(now))
+        .sorted(Comparator
+            .comparing(Curriculum::getCurriculumSubType)
+            .reversed()
+            .thenComparing(Curriculum::getCurriculumSpecialtyCode)
+            .reversed()
+        )
+        .toList();
+  }
+
+  /**
+   * Check whether the given personal details excludes NTN generation.
+   *
+   * @param personalDetails The personal details to check.
+   * @return true if NTN generated cannot continue, else false.
+   */
+  private boolean isExcluded(PersonalDetails personalDetails) {
+    if (personalDetails == null) {
+      log.info("Skipping NTN population as personal details not available.");
+      return true;
+    }
+
+    String gmcNumber = personalDetails.getGmcNumber();
+
+    if (gmcNumber == null || !gmcNumber.matches("\\d{7}")) {
+      String gdcNumber = personalDetails.getGdcNumber();
+
+      if (gdcNumber == null || !gdcNumber.matches("\\d{5}.*")) {
```

Comment:
So all the ones beginning with 'TR' or 'D' (e.g. D11890) are not valid?

Reply:
🤷🏻 

Current SQL:

```sql
case
  when rcm1.GMCNumber LIKE '[0-9][0-9][0-9][0-9][0-9][0-9][0-9]' then rcm1.GMCNumber
  when (rcm1.GMCNumber not LIKE '[0-9][0-9][0-9][0-9][0-9][0-9][0-9]' or rcm1.GMCNumber is null) and rcm1.GDCNumber LIKE '[0-9][0-9][0-9][0-9][0-9]%' then rcm1.GDCNumber
  else 'Invalid_Reference_Number'
end as ReferenceNumber
```

Reply:
Ok, I see these defined in the latest SQL